Libraries/xcexecution/Sources/SimpleExecutor.cpp

```diff
@@ -245,39 +244,66 @@ performInvocations(
                 }
             }
 
-            process::MemoryContext context = process::MemoryContext(
-                invocation.executable().path(),
-                invocation.workingDirectory(),
-                invocation.arguments(),
-                invocation.environment(),
-                processContext->userID(),
-                processContext->groupID(),
-                processContext->userName(),
-                processContext->groupName());
-
-            if (!invocation.executable().builtin().empty()) {
-                /* For built-in tools, run them in-process. */
-                std::shared_ptr<builtin::Driver> driver = _builtins.driver(invocation.executable().builtin());
-                if (driver == nullptr) {
-                    xcformatter::Formatter::Print(_formatter->finishInvocation(invocation, invocation.executable().displayName(), createProductStructure));
+            if (ext::optional<std::string> const &builtin = executable.builtin()) {
+                /* Builtin tool, find and run in-process. */
+                if (std::shared_ptr<builtin::Driver> driver = _builtins.driver(*builtin)) {
+                    xcformatter::Formatter::Print(_formatter->beginInvocation(invocation, *builtin, createProductStructure));
+
+                    process::MemoryContext context = process::MemoryContext(
+                        *builtin,
+                        invocation.workingDirectory(),
+                        invocation.arguments(),
+                        invocation.environment(),
+                        processContext->userID(),
+                        processContext->groupID(),
+                        processContext->userName(),
+                        processContext->groupName());
+                    success = driver->run(&context, filesystem) != 0;
+
+                    xcformatter::Formatter::Print(_formatter->finishInvocation(invocation, *builtin, createProductStructure));
+                } else {
+                    /* Failed to find builtin tool. */
                     return std::make_pair(false, std::vector<pbxbuild::Tool::Invocation>({ invocation }));
                 }
+            } else if (ext::optional<std::string> const &external = executable.external()) {
+                /* External tool, find on the filesystem. */
+                ext::optional<std::string> path;
+                if (FSUtil::IsAbsolutePath(*external)) {
```

Comment:
Yeah, this is the old `Determine()` logic duplicated. :(

I thought about putting it on `Invocation::Executable` itself, but it would have to take a filesystem parameter and it'd be a bit strange to take that on a pure-data class... maybe could do a shared function in this library?

And there's also the issue that the `builtin` branch is different between the two: in here, it looks it up in the builtin registry, but the Ninja executor uses the filesystem. That makes it a little hard to abstract.

Any ideas?